src/gpu/generic/sycl/ref_reduction.cpp

```diff
@@ -0,0 +1,496 @@
+    } else {
+        multi_reduction_ = true;
+        CHECK(init_scratchpad());
+    }
+
+    if (init_conf.stride_size == 1) {
+        init_conf.transpose = false;
+        init_conf.bank_offset = false;
+    } else if (init_conf.stride_size > 4) {
+        init_conf.transpose = true;
+        init_conf.bank_offset = true;
+    } else {
+        init_conf.transpose = false;
+        init_conf.bank_offset = true;
+    }
+
+    auto dims = squeezed_dims_;
+    for (size_t red_iter = 0; red_iter < num_reductions_; ++red_iter) {
+        auto conf = init_conf;
+        const auto &axes = squeezed_axes_;
+        auto dims_begin = dims.begin();
+        auto dims_end = dims.end();
+        auto axis = axes[red_iter];
+        conf.is_first_iter = (red_iter == 0);
+        conf.is_last_iter = (red_iter == num_reductions_ - 1);
+        conf.batch_size = std::accumulate(
+                dims_begin, dims_begin + axis, 1, std::multiplies<int>());
+        conf.reduce_size = dims[axis];
+        if (axis < static_cast<int>(dims.size() - 1)) {
+            conf.stride_size = std::accumulate(
+                    dims_begin + axis + 1, dims_end, 1, std::multiplies<int>());
+        } else {
+            conf.stride_size = 1;
+        }
+
+        needs_atomic_reduction_ = conf.batch_groups != -1;
+        const auto batch_groups
+                = conf.batch_groups == -1 ? conf.batch_size : conf.batch_groups;
+        const auto max_wg_size = max_wg_size_;
+        const auto max_sg_size = max_sg_size_;
+        int tile_col = get_max_col_tile_dim(
+                conf.stride_size, max_wg_size, max_sg_size);
+        tile_col = std::min(tile_col, sycl_reduction_conf_t::local_col_wg);
+        int tile_row = get_max_row_tile_dim(
+                conf.reduce_size, tile_col, max_wg_size, max_sg_size);
+        tile_row = std::min(tile_row, sycl_reduction_conf_t::local_row_wg);
+        conf.tile_col = tile_col;
+        conf.tile_row = tile_row;
+        local_ranges_.emplace_back(range_t {1, tile_row, tile_col});
+
+        auto global_col
+                = round_up_to_nearest_multiple(conf.stride_size, tile_col);
+        auto global_row
+                = round_up_to_nearest_multiple(conf.reduce_size, tile_row);
+        global_ranges_.emplace_back(
+                range_t {conf.batch_size, global_row, global_col});
+        needs_atomic_reduction_ = needs_atomic_reduction_
+                || (global_row > std::min(
+                            tile_row, conf.num_sg_reductions * max_sg_size));
+
+        if (!IMPLICATION(needs_atomic_reduction_, supports_atomics)) {
+            return status::unimplemented;
+        }
+        // Atomics are only supported when deterministic == false
+        if (!IMPLICATION(needs_atomic_reduction_, !attr()->deterministic_)) {
+            return status::unimplemented;
+        }
+        if (needs_atomic_reduction_ && conf.alg == dnnl_reduction_mul) {
+            return status::unimplemented;
```

Comment:
(minor) for these kind of checks, there is a `VDISPATCH` macro that allows to print diagnostics with ONEDNN_VERBOSE mode (see an example [here](https://github.com/oneapi-src/oneDNN/blob/0bff5bd0f56dcf5e68303198996e15a065ffb16d/src/gpu/generic/sycl/ref_convolution.hpp#L91)).

Reply:
Done, thanks.

Reply:
I pushed the change for him as Svet is on holiday